adapters/zerolog/zerolog.go

```diff
@@ -0,0 +1,286 @@
+	}
+
+	// Apply supplied options.
+	for _, option := range opts {
+		if option == nil {
+			continue
+		}
+		option.apply(w)
+	}
+
+	if len(w.levels) == 0 {
+		for _, level := range []zerolog.Level{zerolog.InfoLevel, zerolog.WarnLevel, zerolog.ErrorLevel, zerolog.FatalLevel, zerolog.PanicLevel} {
+			w.levels[level] = struct{}{}
+		}
+	}
+
+	// Create client, if not set.
+	if w.client == nil {
+		var err error
+		if w.client, err = axiom.NewClient(w.clientOptions...); err != nil {
+			return nil, err
+		}
+	}
+
+	// When the dataset name is not set, use "AXIOM_DATASET".
+	if w.datasetName == "" {
+		w.datasetName = os.Getenv("AXIOM_DATASET")
+		if w.datasetName == "" {
+			return nil, ErrMissingDatasetName
+		}
+	}
+
+	go w.runBackgroundJob()
+	return w, nil
+}
+
+func (w *WriteCloser) runBackgroundJob() {
+	var (
+		counter = 0
+		buffer  = &bytes.Buffer{}
+		t       = time.NewTicker(flushInterval)
+		encoder = axiom.ZstdEncoder()
+	)
+
+	flush := func() error {
+		defer func() {
+			counter = 0
+			t.Reset(flushInterval)
+			buffer.Reset()
+		}()
+
+		if buffer.Len() == 0 {
+			return nil
+		}
+
+		r, err := encoder(buffer)
+		if err != nil {
+			return err
+		}
+
+		res, err := w.client.Ingest(w.ctx, w.datasetName, r, axiom.NDJSON, axiom.Zstd, w.ingestOptions...)
+		if err != nil {
+			return fmt.Errorf("failed to ingest events: %w", err)
+		}
+		if res.Failed > 0 {
+			return fmt.Errorf("failed to ingest events: %w", err)
+		}
+		if res.Failed > 0 {
+			logger.Printf("event(s) [%v] at %s failed to ingest: %s\n", res.Failed, res.Failures[0].Timestamp, res.Failures[0].Error)
+			return nil
+		}
+		return nil
+	}
+
+	for {
+		select {
+		case <-w.ctx.Done():
+			if err := flush(); err != nil {
+				logger.Printf("failed to ingest events: %s\n", err)
+			}
+			close(w.closeCh)
+			return
+		case data, ok := <-w.byteCh:
+			if !ok {
+				if err := flush(); err != nil {
+					logger.Printf("failed to ingest events: %s\n", err)
+				}
+				close(w.closeCh)
+				return
+			}
+			if len(data) == 0 {
+				continue
+			}
+
+			counter++
+
+			lvlStr, err := jsonparser.GetUnsafeString(data, zerolog.LevelFieldName)
+			if err != nil {
+				logger.Printf("failed to retrieve level field name from data: %s\n", err)
+				continue
+			}
+
+			lvl, err := zerolog.ParseLevel(lvlStr)
+			if err != nil {
+				logger.Printf("failed to parse level: %s\n", err)
+				continue
+			}
+
+			if _, enabled := w.levels[lvl]; !enabled {
+				continue
+			}
+
+			timeStr, err := jsonparser.GetUnsafeString(data, zerolog.TimestampFieldName)
+			if err != nil {
+				// if Timestamp not set on the logger, default to time.Now()
+				timeStr = zerolog.TimestampFunc().Format(zerolog.TimeFieldFormat)
+			}
+
+			// ignore any errors here
+			data, _ = jsonparser.Set(data, []byte("\""+timeStr+"\""), ingest.TimestampField)
+			data, _ = jsonparser.Set(data, []byte("\""+lvlStr+"\""), zerolog.LevelFieldName)
+			data, _ = jsonparser.Set(data, loggerName, "logger")
```

Comment:
I wonder what's going on here. Isn't the timestamp guaranteed to be there when logged from `zerolog`?

Also there is no need to rewrite the timestamp field, no? We can just tell the Axiom client the name of the timestamp field as well as the timestamp format: https://github.com/axiomhq/axiom-go/blob/main/axiom/ingest/options.go#L30-L41.